Fix Windows EH IP2State tables (remove +1 bias)

sivadeilra opened the pull request:
This changes how LLVM constructs certain data structures that relate to exception handling (EH) on Windows. Specifically this changes how IP2State tables for functions are constructed.  The purpose of this change is to align LLVM to the requires of the Windows AMD64 ABI, which requires that the IP2State table entries point to the boundaries between instructions.

On most Windows platforms (AMD64, ARM64, ARM32, IA64, but *not* x86-32), exception handling works by looking up instruction pointers in lookup tables. These lookup tables are stored in `.xdata` sections in executables. One element of the lookup tables are the `IP2State` tables (Instruction Pointer to State).

If a function has any instructions that require cleanup during exception unwinding, then it will have an IP2State table. Each entry in the IP2State table describes a range of bytes in the function's instruction stream, and associates an "EH state number" with that range of instructions. A value of -1 means "the null state", which does not require any code to execute. A value other than -1 is an index into the State table.

The entries in the IP2State table contain byte offsets within the instruction stream of the function. The Windows ABI requires that these offsets are aligned to instruction boundaries; they are not permitted to point to a byte that is not the first byte of an instruction.

Unfortunately, CALL instructions present a problem during unwinding. CALL instructions push the address of the instruction after the CALL instruction, so that execution can resume after the CALL. If the CALL is the last instruction within an IP2State region, then the return address (on the stack) points to the *next* IP2State region. This means that the unwinder will use the wrong cleanup funclet during unwinding.

To fix this problem, compilers should insert a NOP after a CALL instruction, if the CALL instruction is the last instruction within an IP2State region. The NOP is placed within the same IP2State region as the CALL, so that the return address points to the NOP and the unwinder will locate the correct region.

This PR modifies LLVM so that it inserts NOP instructions after CALL instructions, when needed. In performance tests, the NOP has no detectable significance. The NOP is rarely inserted, since it is only inserted when the CALL is the last instruction before an IP2State transition or the CALL is the last instruction before the function epilogue.

NOP padding is only necessary on Windows AMD64 targets. On ARM64 and ARM32, instructions have a fixed size so the unwinder knows how to "back up" by one instruction.

Interaction with Import Call Optimization (ICO):

Import Call Optimization (ICO) is a compiler + OS feature on Windows which improves the performance and security of DLL imports. ICO relies on using a specific CALL idiom that can be replaced by the OS DLL loader. This removes a load and indirect CALL and replaces it with a single direct CALL.

To achieve this, ICO also inserts NOPs after the CALL instruction. If the end of the CALL is aligned with an EH state transition, we *also* insert a single-byte NOP.  **Both forms of NOPs must be preserved.**  They cannot be combined into a single larger NOP; nor can the second NOP be removed.

This is necessary because, if ICO is active and the call site is modified by the loader, the loader will end up overwriting the NOPs that were inserted for ICO. That means that those NOPs cannot be used for the correct termination of the exception handling region (the IP2State transition), so we still need an additional NOP instruction.  The NOPs cannot be combined into a longer NOP (which is ordinarily desirable) because then ICO would split one instruction, producing a malformed instruction after the ICO call.

sivadeilra committed:
Fix IP2State tables

style: revert one change

Adjust tests

adjust another test

fix llvm/test/CodeGen/XCore/exception.ll

PR feedback

PR feedback

update microsoft-abi-eh-async.cpp for tighter NOP handling

PR feedback: use INT3 instead of NOP for noreturn

update win64_frame.ll

Fix bug, plus more accurate NOP optimization

Fix win64-seh-epilogue-statepoint.ll

sivadeilra commented:
Investigated remaining tests and fixed.

efriedma-quic commented on llvm/test/CodeGen/X86/win64_frame.ll:
```diff
@@ -101,7 +101,8 @@ define void @f5() "frame-pointer"="all" {
 ; CHECK-NEXT:    callq external
-; CHECK-NEXT:    nop
+; UEFI does not have SEH, so does not require NOP
+; WIN64-NEXT:    nop
```
For files with a "Assertions have been autogenerated by utils/update_llc_test_checks.py" comment, please make sure you update them using update_llc_test_checks.py .  Otherwise the next person to modify the file will be confused trying to figure out what you did.

sivadeilra replied:
Happy to do that -- I'm just not familiar with updating those kind of tests yet.  In this case, this test is run 4 times, with different parameters.  Will `update_llc_test_checks.py` know how to do the right thing?  The NOP will only be present in 2 out of the 4 tests.  What should the ideal patch look like, here?

efriedma-quic replied:
update_llc_test_checks.py understands the --check-prefixes flag; it will try to do something reasonable.

efriedma-quic commented on llvm/lib/Target/X86/X86MCInstLower.cpp:
```diff
@@ -2660,6 +2644,133 @@ void X86AsmPrinter::emitCallInstruction(const llvm::MCInst &MCI) {
+    // If there is an EH_LABEL after this CALL, then there is an EH state
+    // transition after this CALL. This is exactly the situation which requires
+    // NOP padding.
+    if (NextMI.isEHLabel()) {
```
As far as I can tell, this triggers on every "invoke"... so if you write `try { f(); f(); f(); } catch (...) {}`, you get a nop after every call.  Not sure how complicated it would be to handle that.

efriedma-quic replied:
At the point where we generate the EH_LABEL instructions, the block layout isn't finalized.  Actually, nothing computes whether there's a state change until computeIP2StateTable() itself.  So maybe the optimization is a little tricky to implement.

efriedma-quic commented on llvm/lib/Target/X86/X86MCInstLower.cpp:
```diff
@@ -2660,6 +2644,133 @@ void X86AsmPrinter::emitCallInstruction(const llvm::MCInst &MCI) {
+// there are no AMD64 instructions that can be encoded in a single byte and
+// which throw C++ exceptions. Unfortunately, this violates the Windows ABI
+// specification, which requires that the IP2State table entries point to the
+// boundaries between exceptions.
```
Does "violating the ABI specification" have any practical side-effects?

sivadeilra replied:
It prevents our analysis tools, such as our hot-patch generator, from correctly analyzing binaries.

namazso replied:
@efriedma-quic this PR supposedly also fixes https://github.com/llvm/llvm-project/issues/54922 which is the same but for SEH. That can and does happen with single byte instructions, as shown in the issue OP.

efriedma-quic commented on clang/test/CodeGenCXX/microsoft-abi-eh-async.cpp:
```diff
@@ -0,0 +1,180 @@
+// RUN: %clang_cl -c --target=x86_64-windows-msvc /EHa -O2 /GS- \
+// RUN:   -Xclang=-import-call-optimization \
+// RUN:   /clang:-S /clang:-o- %s 2>&1 \
+// RUN:   | FileCheck %s
```
`// REQUIRES: x86-registered-target`

efriedma-quic commented on llvm/test/CodeGen/X86/win64-seh-epilogue-statepoint.ll:
```diff
@@ -9,7 +9,6 @@ define i32 @foobar() gc "statepoint-example" personality ptr @__gxx_personality_
 ; CHECK-NEXT:    .seh_endprologue
 ; CHECK-NEXT:    callq bar
 ; CHECK-NEXT:  .Ltmp0:
-; CHECK-NEXT:    nop
```
This makes the call adjacent to the epilog; is that okay?

sivadeilra replied:
Since this is for the Windows GNU target, which does not use SEH, I _believe_ so, but I would really like to hear from someone with experience with the Windows GNU target.

efriedma-quic replied:
Is the codegen change here specific to Windows-GNU targets?  I don't see any reason MSVC targets would behave differently.

mstorsjo replied:
The Windows GNU target does use SEH for the unwind info, but uses GCC compatible unwind info for the language level things (corresponding to ip2state tables I guess).

efriedma-quic replied:
Looking a bit more, if you drop the `[[noreturn]]` and the opt-bisect-limit flag, you get the same result.

sivadeilra replied:
I've reverted this change, because now I'm handling inserting the NOP (before the SEH epilogue, not for all CALLs) when using GNU EH on Win32.

rnk commented:
> The entries in the IP2State table contain byte offsets within the instruction stream of the function. The Windows ABI requires that these offsets are aligned to instruction boundaries; they are not permitted to point to a byte that is not the first byte of an instruction.
> ...
> Unfortunately, this violates the Windows ABI specification, which requires that the IP2State table entries point to the boundaries between exceptions.

I'm a little defensive, since I think I came up with the label+1 scheme, but I don't believe there are any documented requirements covering the ip2state table. I do not believe I have ever seen a description of the rules for ip2state tables anywhere in any Microsoft docs. They appear to be implementation details of MSVC EH personality functions, and the requirements are simply that "exception handling works", and relevant [EH test suites](https://github.com/microsoft/compiler-tests) pass.

I'm very willing to accept that this is a requirement, but I would prefer to change the language in the commit message and comments to speak more about the functional reasons why we should increase code size to get instruction-aligned label offsets, or have references to real, existing, public specifications, instead of references to an undocumented, notional "Windows ABI specification".

IIRC users of IDA Pro have complained, for example, since it attempts to reconstruct EH blocks using these tables, and that's a perfectly good reason to merge a patch to implement this. I think this can be found in the issue tracker if someone looks closely.

> Import Call Optimization (ICO)

I guess text relocations are cool again. :)

sivadeilra commented on llvm/lib/Target/X86/X86MCInstLower.cpp:
```diff
@@ -2660,6 +2644,133 @@ void X86AsmPrinter::emitCallInstruction(const llvm::MCInst &MCI) {
+// termination of the exception handling region (the IP2State transition),
+// so we still need an additional NOP instruction.  The NOPs cannot be combined
+// into a longer NOP (which is ordinarily desirable) because then ICO would
+// split one instruction, producing a malformed instruction after the ICO call.
```
But since the purpose of the larger NOP is to be overwritten during DLL import resolution, that would mean that the final NOP byte would be overwritten, so the IP2State mapping would be incorrect. 

If you mean that an "even larger" NOP could be encoded that way, then I suppose that's right, but I don't see any utility in that.  It would be confusing to debug, too.

namazso commented:
While I haven't checked how exactly this PR interacts with SEH, just here because https://github.com/llvm/llvm-project/issues/54922#issuecomment-2984928111 , I'd like to point out that  https://clang.llvm.org/docs/MSVCCompatibility.html has the following snippet for SEH:

> Asynchronous Exceptions (SEH): Partial. Structured exceptions (__try / __except / __finally) mostly work on x86 and x64. LLVM does not model asynchronous exceptions, so it is currently impossible to catch an asynchronous exception generated in the same frame as the catching __try.

If this patch will remove the +1 offset from the beginning of the call instruction as well, then if that throws it will be caught in the same frame, which is "impossible" (not sure what the practical implications of this are). Call can throw when calling indirect from invalid memory. As it stands, this patch breaks this "functionality", as now these exceptions will be caught in the inner `__try`.

While MSVC never worked this way, this would not fix the whole SEH discrepancy, while effectively "removing" a feature.

namazso commented:
Actually, disregard my previous comment. It seems that it's quite a bit more broken than that already and depends on surrounding code. Making it broken in different ways probably won't hurt much.

sivadeilra commented:
My team is interested in fixing / completing the SEH support in LLVM, as well.  I think this PR is consistent with that, since it aligns LLVM's behavior with that of MSVC.

For an SEH exception that occurs within a function due to a CALL (such as a bogus indirect call), the IP should point to the beginning of the CALL instruction, so the IP2State tables should point to the correct region.

@namazso , I'd be happy to improve the state of things for SEH, as well as ensuring that this PR doesn't regress support for GNU's EH support on Windows.  If GNU's C++ EH support is based on SEH, then I think I will need to change this PR so that a CALL aligned to `.seh_startepilogue` always inserts a NOP call.  That's easy to do.  Do you have pointers to any info on GNU C++ exception dispatch?

namazso commented:
> For an SEH exception that occurs within a function due to a CALL (such as a bogus indirect call), the IP should point to the beginning of the CALL instruction, so the IP2State tables should point to the correct region.

Yes, I was just pointing out that same-frame catch isn't supposed to be supported in LLVM per the docs. By that rule, this code:

```
void test1(void(*a)()) {
    __try {
        a();
    } __except(1) {
        printf("throw in call!\n");
    }
}

void test2(void(*a)()) {
    __try {
        test1(a);
    } __except (1) {
        printf("throw from call!\n");
    }
}

void int3() {
    __debugbreak();
}

int main() {
    test2(&int3);
    test2((void(*)())0x8FFFFFFFFFFFFFFF);
    return 0;
}
```

Should result in this:

```
throw in call!
throw from call!
```

Which indeed would probably work like that if the start labels were also offset by 1 all the time. Of course it's way different from MSVC, but it sounded at least *consistent* even if slightly insane. However when I tested it on the current LLVM release, it was not consistent at all already, so that's why I commented to just disregard this observation - it already didn't conform whatever the docs were saying, it's not going to be any more broken that it currently is.

>  I'd be happy to improve the state of things for SEH, as well as ensuring that this PR doesn't regress support for GNU's EH support on Windows. If GNU's C++ EH support is based on SEH, then I think I will need to change this PR so that a CALL aligned to .seh_startepilogue always inserts a NOP call. That's easy to do. Do you have pointers to any info on GNU C++ exception dispatch?

Sorry, no clue on GNU EH, my observations were purely about the current state of SEH and the docs (incorrectly) describing the "rules" on how exactly it is partially supported.

sivadeilra commented:
@rnk - Aside from the comments / text edits, do you have any other suggestions for this PR?

rnk commented:
Sorry for the delay, I'm at a conference this week.

Regarding wording, I would prefer to rephrase these fragments more precisely in the final commit message: "The Windows ABI requires ..." "... the Windows ABI specification, ..."

The Windows ABI is a nebulous concept, and aligning LLVM codegen with MSVC is something we can only chase so far. At some point, we have to tradeoff bug-for-bug MSVC compatibility with other goals like implementation complexity and code size, and it's helpful to motivate the change with more concrete goals.

In this case, LLVM has grown support for non-call exceptions, so the assumptions this +1 byte scheme no longer hold, and that, to me, is a very compelling reason to do this nop insertion. IDA pro compat seems like a good supporting reason, and then it seems there is a long tail of internal tools that need to parse these ip2state tools.

Hopefully that answers your questions. I expect to be busy over the next month, so please proceed with the review without blocking on my input, I do support this work. :)

dpaoliello commented on clang/test/CodeGenCXX/microsoft-abi-eh-async.cpp:
```diff
@@ -0,0 +1,182 @@
+    return 100;
+}
+
+void case_except_simple_call() NO_TAIL
```
Add checks for these test cases?

sivadeilra replied:
Added checks

dpaoliello commented on clang/test/CodeGenCXX/microsoft-abi-eh-async.cpp:
```diff
@@ -0,0 +1,182 @@
+
+// RUN: %clang_cl -c --target=x86_64-windows-msvc /EHa -O2 /GS- \
+// RUN:   -Xclang=-import-call-optimization \
+// RUN:   /clang:-S /clang:-o- %s 2>&1 \
```
Does this need the `--` before the file name to avoid weird command line parsing issues?

sivadeilra replied:
Added `--`

dpaoliello commented on llvm/lib/Target/X86/X86MCInstLower.cpp:
```diff
@@ -2538,26 +2539,6 @@ void X86AsmPrinter::emitInstruction(const MachineInstr *MI) {
 
   case X86::SEH_BeginEpilogue: {
```
Can this be folded into the case statement above, or will they fail on this assert?

sivadeilra replied:
Maybe, but I don't know the invariants well enough to say whether that assert is appropriate for all of the `SEH_*` ops.

dpaoliello commented on llvm/lib/Target/X86/X86AsmPrinter.h:
```diff
@@ -151,6 +151,8 @@ class LLVM_LIBRARY_VISIBILITY X86AsmPrinter : public AsmPrinter {
 
   void emitCallInstruction(const llvm::MCInst &MCI);
+  bool needsNopAfterCallForWindowsEH(const MachineInstr *MI);
+  void emitNopAfterCallForWindowsEH(const MachineInstr *MI);
```
Nit, more accurately called `maybeEmitNopAfterCallForWindowsEH`

sivadeilra replied:
Renamed

dpaoliello reviewed (approved):
A few minor comments, but otherwise LGTM

arsenm commented on llvm/lib/Target/X86/X86MCInstLower.cpp:
```diff
@@ -2660,6 +2644,122 @@ void X86AsmPrinter::emitCallInstruction(const llvm::MCInst &MCI) {
+bool X86AsmPrinter::needsNopAfterCallForWindowsEH(const MachineInstr *MI) {
+  // We only need to insert NOPs after CALLs when targeting Windows on AMD64.
+  // Since this code is already restricted to X86, we just test for Win64.
+  if (!this->Subtarget->isTargetWin64()) {
```
Don't check this through the subtarget, this is a module level propert y

sivadeilra replied:
Ok.  What should that look like?  I looked at a few variants and it wasn't obvious how they were different from using the subtarget.

kimhak168 commented:
SllfjhvmSSllf

On Mon, Jun 30, 2025, 5:31 AM sivadeilra ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In llvm/lib/Target/X86/X86MCInstLower.cpp
> <https://github.com/llvm/llvm-project/pull/144745#discussion_r2173977583>:
>
> > +// To achieve this, ICO also inserts NOPs after the CALL instruction. If the
> +// end of the CALL is aligned with an EH state transition, we *also* insert
> +// a single-byte NOP.  **Both forms of NOPs must be preserved.**  They cannot
> +// be combined into a single larger NOP; nor can the second NOP be removed.
> +//
> +// This is necessary because, if ICO is active and the call site is modified
> +// by the loader, the loader will end up overwriting the NOPs that were inserted
> +// for ICO. That means that those NOPs cannot be used for the correct
> +// termination of the exception handling region (the IP2State transition),
> +// so we still need an additional NOP instruction.  The NOPs cannot be combined
> +// into a longer NOP (which is ordinarily desirable) because then ICO would
> +// split one instruction, producing a malformed instruction after the ICO call.
> +bool X86AsmPrinter::needsNopAfterCallForWindowsEH(const MachineInstr *MI) {
> +  // We only need to insert NOPs after CALLs when targeting Windows on AMD64.
> +  // Since this code is already restricted to X86, we just test for Win64.
> +  if (!this->Subtarget->isTargetWin64()) {
>
> Ok. What should that look like? I looked at a few variants and it wasn't
> obvious how they were different from using the subtarget.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/llvm/llvm-project/pull/144745#discussion_r2173977583>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BMTQUFSDA3KPV7DATAF5SML3GBSM5AVCNFSM6AAAAAB7TTSKNKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDSNRZGU2DMMJZGQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

arsenm commented on llvm/lib/Target/X86/X86MCInstLower.cpp:
```diff
@@ -2660,6 +2644,122 @@ void X86AsmPrinter::emitCallInstruction(const llvm::MCInst &MCI) {
+bool X86AsmPrinter::needsNopAfterCallForWindowsEH(const MachineInstr *MI) {
+  // We only need to insert NOPs after CALLs when targeting Windows on AMD64.
+  // Since this code is already restricted to X86, we just test for Win64.
+  if (!this->Subtarget->isTargetWin64()) {
```
```suggestion
  const Triple &TT = TM.getTargetTriple();
  if (!TT.isOSWindows() || !TT.getArch() != Triple::x86_64)
```
It's not different, checking module level properties through the subtarget is just an antipattern I've been trying to fix (e.g. 874a02f05b6ebb4b5dbe0ab09beb9c3d5b36e237).

sivadeilra replied:
Updated.

sivadeilra commented:
> Is full 32-bit SEH support ever coming?

Do you mean for 32-bit x86?  It's not a focus of mine, currently.

void2012 commented:
> Do you mean for 32-bit x86? 
Yes.

> It's not a focus of mine, currently.
Sad. I wish I could figure out LLVM code and contribute too. How do you learn things like that?

sivadeilra commented:
@arsenm, @efriedma-quic, @namazso , do you have more feedback on this PR?  I'd like to get this merged and move on to the next pieces for supporting Windows hot-patching.

sivadeilra commented:
@arsenm, @efriedma-quic , @namazso - Ping.  It's been more than a week since I've gotten any feedback on this PR.  Can someone take a look at this please?

efriedma-quic commented on llvm/lib/Target/X86/X86MCInstLower.cpp:
```diff
@@ -2537,26 +2538,6 @@ void X86AsmPrinter::emitInstruction(const MachineInstr *MI) {
-    // Check if preceded by a call and emit nop if so.
-    for (MBBI = PrevCrossBBInst(MBBI);
-         MBBI != MachineBasicBlock::const_iterator();
-         MBBI = PrevCrossBBInst(MBBI)) {
```
I'm seeing a warning "unused function PrevCrossBBInst"

sivadeilra replied:
Thanks; deleted the dead func.

sivadeilra commented:
> You end up in the wrong region if g1 unwinds, I think.

I see.  I've reproduced it with the input and args you provided; thank you.

I think the conservative change would be to insert a NOP for a CALL that is at the end of an MBB.  Making that change and re-running, I see that it does insert the NOP after the CALL to `g1`.  For a `[[noreturn]]` function that seems fine; we'll never run the NOP so we don't even need to decode it, and the only price is a tiny bit of binary size increase.  However, if that patterns is repeated often enough, such as panic handlers in Rust (since there are many of them inserted into code, even if they rarely run), then the cost could become substantial.

What would you recommend?

sivadeilra commented:
That was easier than expected -- thanks again, @efriedma-quic .  I've updated the PR.  I've also avoided adding NOPs after tail calls, which were happening in my first version of this.  There's never a need for NOP padding after TAILJMP* because there won't be a return address pointing into this frame.

efriedma-quic commented:
For the cases involving noreturn functions, can we emit int3 instead of nop?  (You can tell a call is noreturn if it's the last instruction in a basic block, and that basic block has no successors.)  Compare to X86AvoidTrailingCallPass, which already inserts int3 instructions.

sivadeilra commented:
> For the cases involving noreturn functions, can we emit int3 instead of nop?

Sure.  I've implemented that, verified that tests pass, and updated the PR.  I also added coverage for this to `noreturn-call-win64.ll`.

efriedma-quic commented on llvm/test/CodeGen/X86/stack-coloring-wineh.ll:
```diff
@@ -84,12 +84,12 @@ define void @pr66984(ptr %arg) personality ptr @__CxxFrameHandler3 {
 ; X86_64-NEXT:    movq %rcx, {{[-0-9]+}}(%r{{[sb]}}p) # 8-byte Spill
 ; X86_64-NEXT:  .Ltmp0:
 ; X86_64-NEXT:    callq throw
+; X86_64-NEXT:    nop
```
I think this is optimized code.

I didn't really think that deeply when I suggested looking for a block with no successors.  In this case, you end up with two edges: an unwind edge, and a fallthrough edge to a an empty block with no successors.  Which is a pattern you could detect, I guess.

I don't want the initial patch here to get too complicated here, though, so it would be fine to just file an issue for this, and look into it later.

efriedma-quic commented on llvm/test/CodeGen/X86/taildup-heapallocsite.ll:
```diff
@@ -37,9 +37,11 @@ cond.end:                                         ; preds = %entry, %cond.true
 ; CHECK: testq
 ; CHECK: je
 ; CHECK: callq alloc
+; CHECK-NEXT: nop
```
There is no handler?  This function doesn't contain any exception-handling constructs at all.

MSVC does not insert a nop before the epilogue in this case, as far as I can tell.

sivadeilra replied:
From my testing, MSVC actually does insert a NOP before the SEH epilogue.  The tricky part is convincing MSVC to avoid tail-call optimization.  Compiling with optimizations disabled, I see a CALL / NOP / epilogue sequence in this code, compiled as C with exceptions disabled.  Same results with /EHa.

```
void foo(int);
void bar() {
  foo(42);
}
```

sivadeilra replied:
Ah, thank you for the example, I just hadn't noticed that this was updated.

I'm actually a little surprised that MSVC isn't inserting a NOP in this situation. In my own test, I was relying on disabling optimizations in order to disable the tail-call, but apparently that's having more effects that I intended.  So your example helps clarify this.

It seems like the rule is "If a CALL is followed by `SEH_BeginEpilogue` *and* the CALL is not inside a non-null EH state, then no NOP is required".  Since the transition from a non-null EH state to a null state would be indicated by an `EH_LABEL`, it's equivalent to "if a CALL is followed by `SEH_BeginEpilogue` then no NOP is required".  Is that consistent with your understanding?

If so, then I can simplify the logic that handles `SEH_BeginEpilogue` so that it just returns.  Does that reasoning sound right to you?

sivadeilra replied:
I went further with more precise optimization around NOP insertion.  We should only be inserting NOPs when a function has an EH personality function (and thus has IP2State tables).  No IP2State table mean no need to insert NOPs at all, even for SEH, right?  The `WinException::beginFunction` computes `shouldEmitPersonality`, so I used that value to control whether we do any NOP insertion.

Some tests now fail due to differences.  I noticed in `llvm\test\CodeGen\X86\win64_call_epi.ll` that `foo1` looks correct (and still passes), but `foo2` appears to be incorrect.  `foo2` does not have an EH personality function, but it checks that ` nop` is inserted after `callq`.  After my changes (for using `shouldEmitPersonality`), the test for `foo2` fails.

What's the right thing to do here?  Should `foo2` be using `personality ptr @personality` and should its `call` be upgraded to `invoke`, or should `foo2` not expect to find a `nop`?

sivadeilra replied:
I implemented the more-accurate optimizations for NOP padding (only inserting when we're also building IP2State tables).  It had the benefit of removing NOPs from a lot of codegen situations that were not even using EH.  I've pushed an update to the PR as a separate commit, which should make it a bit easier to review.

sivadeilra replied:
I checked MSVC, and when compiling with both SEH (/EHa) and C++ exceptions, but with no handlers in the current function, there is no NOP before the SEH epilogue.  Which makes sense -- if you're unwinding a stack and a frame has nothing to do, it doesn't matter that IP points to the epilogue.  The unwind tables have unwind codes specifically for that situation, which are unrelated to running handlers.

I think for `win64_call_epi.ll`, the main thing is to enable the EH personality function for `foo2`.  There are plenty of others tests (such as the ones updated in this PR now) that implicitly verify that there is no NOP before the epilogue (when there is no EH personality function), so that's covered.

efriedma-quic replied:
My point was just that nop insertion shouldn't specifically check for whether we're using the C++ personality function, as opposed to the SEH personality function; both personality functions require nop insertion.  Just wanted to point that out because you mentioned IP2State tables in particular.

Your plan for win64_call_epi.ll seems fine.

sivadeilra replied:
Sounds good.  In this case, I'm not checking for a specific EH personality function, I'm just using the same logic that controls whether we run `computeIP2State`.

Looks like an unrelated Clang break occurred in the CI runs.  I've rebased my PR and verified that the same Clang test failure is fixed, after the rebase.

Provided the CI jobs are green, is there anything else needed in this PR?

namazso commented:
Not sure how relevant this is, but EDK2 by default will strip all exception data: https://github.com/tianocore/edk2/blob/a1b509c1a453815acbc6c8b0fc5882fd03a6f2c0/BaseTools/Source/C/GenFw/GenFw.c#L2447

Although there is an option to GenFw to keep them. Performance impact of the nops is negligible, so maybe it's a better choice to just keep those. After all, MSVC also generates and keeps them (since it's entirely unaware of UEFI), and it's just GenFw removing the directory and section.

sivadeilra commented:
I've updated `win64_frame.ll`.

For UEFI, it seems unimportant.  We could use `WinException::shouldEmitPersonality` to control whether any NOPs get inserted at all, since that's the same condition that governs whether we call `computeIP2StateTable`.

frobtech commented:
> Actually, thinking about it, maybe we should just do the nop insertion for UEFI targets. The UEFI environment is very similar to Windows. CC @Prabhuk .

I think that's an appropriately conservative place to start, anyway.  The uefi-* targets are still nascent and heretofore the Windows targets have been used for producing UEFI binaries.  So the approach we've taken is that for all miscellaneous code generation issues, we should start with the UEFI target matching the behavior of the corresponding Windows target exactly.  That seems like the right thing to do for now.  In the future as the uefi-* targets mature, we can reconsider things that are really only relevant to Windows ABIs and not to the UEFI ABI or the PE-COFF format.  The Windows EH ABI is not an aspect of the UEFI ABI, but the support for Windows EH does not interfere with UEFI so I think it's harmless enough to emit all the same stuff.

efriedma-quic reviewed (approved):
LGTM

The following might warrant a followup at some point:

- generating int3 instead of nop after "always-throw" calls, like stack-coloring-wineh.ll
- elide nops between two calls if they have the same catch handler.
- elide the nop between a call and the epilogue if the call is in a "null" region, i.e. the exception won't be handled.  (Not 100% sure about this, but it seems like MSVC does it; https://godbolt.org/z/3fs97Pe4s .)

Please file issues for those so we have a record of that.  But I'm happy to land this without those optimizations.

sivadeilra commented:
Created issue #150057 for follow-up improvements.

sivadeilra commented:
Looking at this now, along with a test failure on AArch64.

dyung commented:
> This is causing failures on my macOS machine that uses paths starting with "/Users":

These test failures also appeared on my macOS bot: https://lab.llvm.org/buildbot/#/builders/190/builds/23983

sivadeilra commented:
I had a fix PR going through CI jobs, but 9a7a6b2 was committed before my job finished.  It added `--`, which I think is necessary but not sufficient.  I'll revive my fix PR, which also changed `/foo` to `-foo` for flags.

dyung commented:
> > We are still seeing the above failures downstream on Linux, even with the fix in [9a7a6b2](https://github.com/llvm/llvm-project/commit/9a7a6b24f8c88718ef1eea57bed07332e751c027)
> 
> Merged #150145. Let me know if you still see errors after that.

The 2 tests are still failing: https://lab.llvm.org/buildbot/#/builders/190/builds/24008

Any other ideas or should we just revert everything while you try to figure it out?

dpaoliello commented:
> Do folks on this PR have any ideas on what might be going wrong? For example, does Rust need to adapt its llvm codegen for this in some way? Or does the test do something that looks bad?

If you're using inline asm to do a call, then you'll also need to manually add the `nop` that this PR inserts automatically.

rnk commented:
I got as far as prototyping this quick fix, but I don't have a test case, and I have to head home soon for the day, if anyone wants to pick it up:
```
❯ git diff
diff --git a/llvm/lib/Target/X86/X86MCInstLower.cpp b/llvm/lib/Target/X86/X86MCInstLower.cpp
index 481a9be8374a..8f5fdbadab57 100644
--- a/llvm/lib/Target/X86/X86MCInstLower.cpp
+++ b/llvm/lib/Target/X86/X86MCInstLower.cpp
@@ -2612,8 +2612,12 @@ void X86AsmPrinter::emitInstruction(const MachineInstr *MI) {
   MCInst TmpInst;
   MCInstLowering.Lower(MI, TmpInst);

-  if (MI->isCall()) {
-    emitCallInstruction(TmpInst);
+  if (MI->isCall() || MI->isInlineAsm()) {
+    if (MI->isCall())
+      emitCallInstruction(TmpInst);
+    else
+      EmitAndCountInstruction(TmpInst);
+
     // Since tail calls transfer control without leaving a stack frame, there is
     // never a need for NOP padding tail calls.
     if (!IsTailJump)
```

sivadeilra commented:
Ironically, the motivation for this PR was to improve Clang and Rust on Windows.

Can you provide more detail on the Rust failure?  I will help diagnose ASAP.

zmodem commented:
> Maybe we should make the compiler insert a nop after inline asm blocks marked "unwind"? Shifting the burden to the user seems error-prone, and the extra nop is unlikely to matter much.

+1 this sounds like a good way to unbreak users.

sivadeilra commented:
Are you proposing adding a `nop` after every inline asm block if it is in a block marked with unwind semantics?  That might be too expensive.  It might needlessly impact code such as vector graphics / vector matrix multiply.

Can you insert a NOP only if the last instruction in the block is a CALL?

efriedma-quic commented:
Not every inline asm, only inline asm statements that specifically have an "unwind" clobber (or in Rust, may_unwind).  Which is pretty rare, in practice: I've never seen anyone use it in clang, and it isn't even available on stable in Rust.

Accurately detecting whether the last instruction of an inline asm block is a call would require an assembler directive which would be extremely nasty to implement correctly, because it interacts with other assembler features like alignment directives.

dpaoliello commented:
@zmodem can you please file an issue so that this doesn't get dropped. Not sure who's going to take care of fixing it...
